bumble/profiles/vocs.py

```diff
@@ -0,0 +1,244 @@
+    ) -> None:
+
+        self.volume_offset = volume_offset
+        self.change_counter = change_counter
+        self.audio_channel_location = audio_channel_location
+        self.audio_output_description_value = audio_output_description_value
+
+        characteristics = []
+
+        self.volume_offset_state = gatt.Characteristic(
+            uuid=gatt.GATT_VOLUME_OFFSET_STATE_CHARACTERISTIC,
+            properties=(
+                gatt.Characteristic.Properties.READ
+                | gatt.Characteristic.Properties.NOTIFY
+            ),
+            permissions=gatt.Characteristic.Permissions.READ_REQUIRES_ENCRYPTION,
+            value=gatt.CharacteristicValue(read=self._on_read_volume_offset_state),
+        )
+        characteristics.append(self.volume_offset_state)
+
+        self.audio_location = gatt.Characteristic(
+            uuid=gatt.GATT_AUDIO_LOCATION_CHARACTERISTIC,
+            properties=(
+                gatt.Characteristic.Properties.READ
+                | gatt.Characteristic.Properties.NOTIFY
+                | gatt.Characteristic.Properties.WRITE_WITHOUT_RESPONSE
+            ),
+            permissions=(
+                gatt.Characteristic.Permissions.READ_REQUIRES_ENCRYPTION
+                | gatt.Characteristic.Permissions.WRITE_REQUIRES_ENCRYPTION
+            ),
+            value=gatt.CharacteristicValue(
+                read=self._on_read_audio_location,
+                write=self._on_write_audio_location,
+            ),
+        )
+        characteristics.append(self.audio_location)
+
+        self.volume_offset_control_point = gatt.Characteristic(
+            uuid=gatt.GATT_VOLUME_OFFSET_CONTROL_POINT_CHARACTERISTIC,
+            properties=gatt.Characteristic.Properties.WRITE,
+            permissions=gatt.Characteristic.Permissions.WRITE_REQUIRES_ENCRYPTION,
+            value=gatt.CharacteristicValue(
+                write=self._on_write_volume_offset_control_point
+            ),
+        )
+        characteristics.append(self.volume_offset_control_point)
+
+        self.audio_output_description = gatt.Characteristic(
+            uuid=gatt.GATT_AUDIO_OUTPUT_DESCRIPTION_CHARACTERISTIC,
+            properties=(
+                gatt.Characteristic.Properties.READ
+                | gatt.Characteristic.Properties.NOTIFY
+                | gatt.Characteristic.Properties.WRITE_WITHOUT_RESPONSE
+            ),
+            permissions=(
+                gatt.Characteristic.Permissions.READ_REQUIRES_ENCRYPTION
+                | gatt.Characteristic.Permissions.WRITE_REQUIRES_ENCRYPTION
+            ),
+            value=gatt.CharacteristicValue(
+                read=self._on_read_audio_output_description,
+                write=self._on_write_audio_output_description,
+            ),
+        )
+        characteristics.append(self.audio_output_description)
+
+        super().__init__(characteristics=characteristics, primary=False)
+
+    @property
+    def volume_offset_state_bytes(self) -> bytes:
```

Comment:
(this comment also applies to the other characteristics in this service)
Instead of having customized bytes serialization methods here, you should have shared adapters, used both for the service and service proxy. 
Look at the `heart_rate_service.py` for an example of how this may be done.

Reply:
I simplified the service implementation and added standalone classes for each characteristic with shared methods as you suggested.